Prose/ProseLib/Sources/ConversationFeature/MessageBar/MessageBarTextField.swift

```diff
@@ -62,42 +69,76 @@ struct MessageBarTextField: View {
 
 // MARK: Reducer
 
+enum MessageBarTextFieldEffectToken: Hashable, CaseIterable {
+  case pauseTyping
+}
+
+public extension DispatchQueue.SchedulerTimeType.Stride {
+  static let pauseTypingDebounceDuration: Self = .seconds(30)
+}
+
 public let messageBarTextFieldReducer: Reducer<
   MessageBarTextFieldState,
   MessageBarTextFieldAction,
-  Void
-> = Reducer { state, action, _ in
+  ConversationEnvironment
+> = Reducer { state, action, environment in
   switch action {
   case .sendTapped where !state.message.isEmpty:
     let content = state.message
     state.message = ""
     return Effect(value: .send(message: content))
 
-  default:
+  case .binding(\.$isFocused) where state.isFocused && !state.message.isEmpty,
```

Comment:
What do you think about this alternative reducer implementation? …

It's not less code, quite the opposite. However when the reducer acts on a change of state I often find it easier to user `onChange(of:)` instead of switching over actions + state.

```swift
public let messageBarTextFieldReducer: Reducer<
  MessageBarTextFieldState,
  MessageBarTextFieldAction,
  ConversationEnvironment
> = Reducer { state, action, environment in
  switch action {
  case .sendTapped where !state.message.isEmpty:
    let content = state.message
    state.message = ""
    return Effect(value: .send(message: content))

  case let .setChatState(chatState):
    return environment.proseClient
      .sendChatState(state.chatId, chatState)
      .fireAndForget()

  case .onDisappear:
    return .cancel(token: MessageBarTextFieldEffectToken.self)

  case .sendTapped, .send, .binding:
    return .none
  }
}
.binding()
.onChange(of: TextFieldState.init) { current, state, action, environment in
  switch (current.isFocused, current.message.isEmpty) {
    // If the textfield is not focused or empty we don't consider the user typing.
    case (false, _), (_, true):
      return .merge([
        environment.proseClient
          .sendChatState(state.chatId, .active)
          .fireAndForget(),
        .cancel(id: MessageBarTextFieldEffectToken.pauseTyping),
      ])
    
    case (true, _), (_, false):
      return .merge([
        environment.proseClient
          .sendChatState(state.chatId, .composing)
          .fireAndForget(),
        Effect(value: .setChatState(.paused))
          .delay(for: .pauseTypingDebounceDuration, scheduler: environment.mainQueue)
          .eraseToEffect()
          .cancellable(id: MessageBarTextFieldEffectToken.pauseTyping, cancelInFlight: true),
      ])
  }
}

private struct TextFieldState: Equatable {
  var message: String
  var isFocused: Bool
  
  init(_ state: MessageBarTextFieldState) {
    self.message = state.message
    self.isFocused = state.isFocused
  }
}
```

Reply:
This is better, you're right. I'm not used to using `.onChange` but I should definitely think about it more often.

I just changed `switch (current.isFocused, current.message.isEmpty)` for `if !current.isFocused || current.message.isEmpty`.